### ISSUE TITLE
Organization: suggestion on the addition of loss and other niceties to Flux

### ISSUE THREAD

**Issue by bhvieira** [FluxML/Flux.jl#826]
In a few PRs (e.g. #818) loss functions are added to Flux, which already implements a few. This is common among frameworks, e.g. Keras supports many loss functions (https://keras.io/losses/).

While there aren't that many on Flux#master right now, the total amount of possible loss functions is enormous, perhaps we could reach a consensus on what to do with loss functions? I fear that these might start bloating Flux (including testing and also the PR and Issues tracks)

I have suggestions: JuliaML's [LossFunctions,jl](https://github.com/JuliaML/LossFunctions.jl) supports most of the common ones and also a few more esoteric types. It could be used (examples and elsewhere), showing that these are implemented already, though I'm not sure about how well LossFunctions+Flux behave together, haven't tested.

Another way would be to simply start a new lightly-maintained repository (e.g. FluxLossFunctions, FluxExtended, FluxUtils) with such useful functions without bloating Flux main repository.

Thoughts?

**Comment by Tokazama** [comment]
LossFunctions.jl definetely has the best documentation but if you want something with a fairly complete set of distance functiions take a look at [Distances.jl](https://github.com/JuliaStats/Distances.jl).

**Comment by bhvieira** [comment]
Fair point @Tokazama.

My idea here is not to "canonize" a library as the 'official' Flux cost/distance function supplier, but to foment discussion on Flux organization regarding this. `layers/stateless.jl` already has `mse`, `binarycrossentropy`, `logitbinarycrossentropy`, and `normalise`. Neither are exported, and I consider these the bare minimum a library should have, but it's fairly obvious to see that, if we keep adding it could possibly never stop.

Now, I'm not saying that's undesirable (in fact, it makes Flux easier to use, specially to novices in Julia).

From my own experience: `mlr` in R has to support many loss functions, see how many they amassed already [here](https://mlr.mlr-org.com/articles/tutorial/measures.html) (disclaimer: I added quite a few of these).

**Comment by bhvieira** [comment]
BTW if the repository owners think it's actually useful to have the most loss functions possible (for example, to achieve feature parity with, say, Keras), then we could move into that direction as well.

**Comment by ToucheSir** [comment]
My preference would be to rely on external libraries wherever possible, but we'd have to figure out what those are. Also, most of them would probably require either ChainRulesCore and/or GPU support. There may be a place for spinning of a separate package from the current losses as well, but thus far I've not seen much clamouring for them outside of Flux users.

**Comment by bhvieira** [comment]
Will close this one since it's been a while and Flux changed a lot in the meantime
